### PR TITLE
Add masked average pooling for pooling with segmentation masks (DetCon)

### PR DESCRIPTION
**Pull request by liopeer**
## Changes
 - add masked pooling for batched and unbatched feature maps together with integer masks

## How has it been tested
 - validated against examples calculated by hand

### CODE REVIEW

File: tests/models/test_ModelUtils.py [hunk 1]
```diff
@@ -6,22 +6,113 @@
...
     batch_unshuffle,
     deactivate_requires_grad,
+    masked_pooling,
     nearest_neighbors,
     normalize_weight,
     update_momentum,
 )
 
+scatter_min_torch_version = version.parse("1.12.0")
+torch_version = version.parse(torch.__version__)
```
Comment (guarin) [inline tests/models/test_ModelUtils.py:30]: Instead of depending on `packaging` we usually check if the feature is available:

```suggestion 
is_scatter_reduce_available = hasattr(Tensor, "scatter_reduce_")
```

File: tests/models/test_ModelUtils.py [hunk 2]
```diff
@@ -6,22 +6,113 @@
...
+        return res
+
+    @pytest.mark.parametrize(
+        "feature_map, mask, expected_result",
+        [
+            ("feature_map1", "mask1", "expected_result1"),
+            ("feature_map2", "mask2", "expected_result2"),
+        ],
+    )
+    def test__mask_reduce(self, feature_map, mask, expected_result, request) -> None:
```
Comment (guarin) [inline tests/models/test_ModelUtils.py:83]: Please also add types in tests, it helps understanding the tests

File: tests/models/test_ModelUtils.py [hunk 3]
```diff
@@ -6,22 +6,113 @@
...
+        feature_map = request.getfixturevalue(feature_map)
+        mask = request.getfixturevalue(mask)
+        expected_result = request.getfixturevalue(expected_result)
```
Comment (guarin) [inline tests/models/test_ModelUtils.py:86]: This test is hard to understand as one has to follow the parametrization and fixtures and figure out how the maps, masks, and results fit together.

I propose to restructure the test to:
```
    @pytest.mark.parametrize(
        "feature_map,  mask, expected_result",
        [
            (
                torch.tensor([[[0, 1], [2, 3]], [[4, 5], [6, 7]], [[8, 9], [10, 11]]], dtype=torch.float32),
                torch.tensor([[0, 0], [1, 2]], dtype=torch.int64),
                torch.tensor([[0.5, 2.0, 3.0], [4.5, 6.0, 7.0], [8.5, 10.0, 11.0]], dtype=torch.float32),
            ),
            (...),
        ],
    )
```

This makes it clearer which values are used for the test and it is easier to understand if the map, mask, and result match.

The goal in tests is really that they should be super easy to understand, otherwise it becomes hard to verify if the test is correctly implemented.

**Review by guarin — COMMENTED** [review]
This looks great! Left some comments regarding formatting and code structure but the logic looks correct

File: lightly/models/utils.py [hunk 4]
```diff
@@ -27,6 +27,75 @@
...
+        return _mask_reduce(source, mask, reduce, num_cls)
+    elif source.dim() == 4:
+        return _mask_reduce_batched(source, mask, num_cls)
+    else:
+        raise ValueError("source must have 3 or 4 dimensions")
+
+
+def _mask_reduce(
+    source: Tensor, mask: Tensor, reduce: str = "mean", num_cls: Optional[int] = None
+) -> Tensor:
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/models/utils.py:59]: I was actually orienting myself at `torch.nn` layers e.g. `Conv2d` and `MaxPool2d` which also support batched and non-batched inputs. However, from what I have seen now, the functional API that they call also only works on batched inputs.

What do you think about kicking the unbatched implementation out, but creating an additional `nn.Module` wrapper `MaskedPooling` that works with unbatched and batched inputs? This would then truly follow the torch conventions.

File: lightly/models/utils.py [hunk 5]
```diff
@@ -27,6 +27,75 @@
     from timm.models.vision_transformer import VisionTransformer
 
 
+def masked_pooling(
+    source: Tensor, mask: Tensor, reduce: str = "mean", num_cls: Optional[int] = None
+) -> Tensor:
+    """Reduce image feature maps (B x C x H x W) or (C x H x W) according to an integer
+    index given by `mask` (B x H x W) or (H x W).
+
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/models/utils.py]: fixed in https://github.com/lightly-ai/lightly/pull/1739/commits/163f752a314db8bc5af1b237084477ccd9c365a3

File: lightly/models/utils.py [hunk 6]
```diff
@@ -27,6 +27,75 @@
...
+    source: Tensor, mask: Tensor, reduce: str = "mean", num_cls: Optional[int] = None
+) -> Tensor:
+    c, h, w = source.shape
+    if num_cls is None:
+        cls = mask.unique(sorted=True)
+    else:
+        cls = torch.arange(num_cls, device=mask.device)
+    num_cls = cls.size(0)
+    # create output tensor
+    output = torch.zeros((c, num_cls), device=source.device)  # (C N)
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/models/utils.py]: neat! fixed in https://github.com/lightly-ai/lightly/pull/1739/commits/ec5ddd2a60acfcce53e18fd6c6e473a82181144b

File: tests/models/test_ModelUtils.py [hunk 7]
```diff
@@ -6,22 +6,113 @@
 import pytest
 import torch
 import torch.nn as nn
+from packaging import version
```
[the remark this answers is not in the document]
Comment (liopeer) [reply tests/models/test_ModelUtils.py]: thanks! fixed in https://github.com/lightly-ai/lightly/pull/1739/commits/298e388ef88fed7005c67179edf1ba60194b5896

File: tests/models/test_ModelUtils.py (hunk 2 above)
Comment (liopeer) [reply tests/models/test_ModelUtils.py]: fixed in https://github.com/lightly-ai/lightly/pull/1739/commits/9a631d84fd0caf4e67d5dc9c841292c7950a4a2b

File: tests/models/test_ModelUtils.py (hunk 3 above)
Comment (liopeer) [reply tests/models/test_ModelUtils.py]: fixed in https://github.com/lightly-ai/lightly/pull/1739/commits/f90d1e6eb948c95c10921b4c2cf46401a822f799

File: lightly/models/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/models/utils.py:59]: kept the unbatched version, but now calling the batched one from within there as you suggested: https://github.com/lightly-ai/lightly/pull/1739/commits/b5ba0b94bb3c00e8fa3b24138287ac512847c93b

File: lightly/models/utils.py [hunk 8]
```diff
@@ -27,6 +27,75 @@
...
+    else:
+        raise ValueError("source must have 3 or 4 dimensions")
+
+
+def _mask_reduce(
+    source: Tensor, mask: Tensor, reduce: str = "mean", num_cls: Optional[int] = None
+) -> Tensor:
+    c, h, w = source.shape
+    if num_cls is None:
+        cls = mask.unique(sorted=True)
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/models/utils.py]: added in https://github.com/lightly-ai/lightly/pull/1739/commits/f90d1e6eb948c95c10921b4c2cf46401a822f799

File: lightly/models/utils.py [hunk 9]
```diff
@@ -27,6 +27,63 @@
     from timm.models.vision_transformer import VisionTransformer
 
 
+def masked_pooling(
```
[the remark this answers is not in the document]
Comment (liopeer) [reply lightly/models/utils.py]: fixed in https://github.com/lightly-ai/lightly/pull/1739/commits/484cdc5a4cf233669d2e1e551b8a9c09ba2d5734
